libkbfs/backpressure_disk_limiter.go

```diff
@@ -15,52 +15,232 @@ import (
 	"golang.org/x/net/context"
 )
 
-// backpressureDiskLimiter is an implementation of diskLimiter that
-// uses backpressure to slow down block puts before they hit the disk
-// limits.
+// backpressureTracker keeps track of the variables used to calculate
+// backpressure. It keeps track of a generic resource (which can be
+// either bytes or files).
 //
-// Let J be the (approximate) byte usage of the journal and F be the
-// free bytes on disk. Then we want to enforce
+// Let U be the (approximate) resource usage of the journal and F be
+// the free resources. Then we want to enforce
 //
-//   J <= min(k(J+F), L),
+//   U <= min(k(U+F), L),
 //
-// where 0 < k <= 1 is some fraction, and L > 0 is the absolute byte
-// usage limit. But in addition to that, we want to set thresholds 0
-// <= m <= M <= 1 such that we apply proportional backpressure (with a
-// given maximum delay) when
+// where 0 < k <= 1 is some fraction, and L > 0 is the absolute
+// resource usage limit. But in addition to that, we want to set
+// thresholds 0 <= m <= M <= 1 such that we apply proportional
+// backpressure (with a given maximum delay) when
 //
-//   m <= max(J/(k(J+F)), J/L) <= M,
+//   m <= max(U/(k(U+F)), J/L) <= M,
 //
 // which is equivalent to
 //
-//   m <= J/min(k(J+F), L) <= M.
-//
-// TODO: Also do limiting based on file counts.
+//   m <= U/min(k(U+F), L) <= M.
+type backpressureTracker struct {
```

Comment:
There are some assumptions about the user of this tracker doing locking themselves, right?  That should be written into the struct comment somewhere.